apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/KmsTest.cs

```diff
@@ -0,0 +1,276 @@
+
+            // We should be able to download it again, even with a default (no CSEK) client.
+            TestHelpers.ValidateData(bucketName, objectName, data);
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasDefaultKmsKey_UploadWithNoSpecifiedKey_DownloadWithClientWithCsek()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string bucketName = CreateBucketWithDefaultKmsKeyName(KeyName1);
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // Upload an object without specifying anything, so it should be encrypted with key 1.
+            Object obj = client.UploadObject(bucketName, objectName, null, data);
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // We should be able to download the file even if we use a client that supplies CSEK by default.
+            EncryptionKey key = EncryptionKey.Generate();
+            var encryptionKeyClient = StorageClient.Create(encryptionKey: key);
+            TestHelpers.ValidateData(bucketName, objectName, data, encryptionKeyClient);
+        }
+
+        [SkippableFact]
+        public void UploadObject_SpecifyBothCsekAndKmsKey()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // It's invalid to specify both an EncryptionKey (that's non-None) and a key name
+            var options = new UploadObjectOptions { EncryptionKey = EncryptionKey.Generate(), KmsKeyName = KeyName1 };
+            var exception = Assert.Throws<ArgumentException>(() => client.UploadObject(_fixture.InitiallyEmptyBucket, objectName, null, data, options));
+            Assert.Contains(nameof(CopyObjectOptions.EncryptionKey), exception.Message);
+            Assert.Contains(nameof(CopyObjectOptions.KmsKeyName), exception.Message);
+        }
+
+        [SkippableFact]
+        public void CopyObject_ChangeKey()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+            string bucketName = _fixture.InitiallyEmptyBucket;
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // Upload an object with a KMS key.
+            Object obj = client.UploadObject(bucketName, objectName, null, data, new UploadObjectOptions { KmsKeyName = KeyName1 });
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // Copy it (within the same bucket, for simplicity) with a different key
+
+            string copyName = objectName + "-copy";
+            Object obj2 = client.CopyObject(bucketName, objectName, bucketName, copyName, new CopyObjectOptions { KmsKeyName = KeyName2 });
```

Comment:
@jskeet this is this a rewrite or a object copy operation (in the underlying request)? `storage.objects.copy` will not support CMEK based on updated documentation in cl/187933097.

Reply:
Yes, it uses rewrite. It seems to work at the moment... will check the doc change you mention.

Reply:
Whoops, had misread your comment. Yes, it uses rewrite so it *is* okay and is expected to be.

Reply:
Thanks for clarifying!